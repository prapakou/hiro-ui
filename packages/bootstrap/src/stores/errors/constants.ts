import { ActionType } from "typesafe-actions";

export const ERROR_NAMESPACE = "error";

export interface ErrorMessage {
  message: string;
  name?: string;
  code?: number;
}

export type ErrorStateType = ErrorMessage | null;
export type ErrorActionsType = ActionType<typeof import("./actions")>;
