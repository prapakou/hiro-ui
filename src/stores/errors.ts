import { BehaviorSubject } from "rxjs";

import { createStateGetter } from "../helpers";

type ErrorState = Error | undefined;

const error$ = new BehaviorSubject<ErrorState>(undefined);

export const errorStore = {
  actions: {
    clearError: () => error$.next(undefined),
    setError: (e: Error) => error$.next(e)
  },
  getters: {
    useError: createStateGetter<ErrorState>(error$)
  }
};
