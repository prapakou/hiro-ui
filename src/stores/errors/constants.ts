import { ActionType } from "typesafe-actions";

export const ERROR_NAMESPACE = "error";

export interface ErrorMessage {
  message: string;
  name?: string;
}

export type ErrorStateType = ErrorMessage | null;
export type ErrorActionsType = ActionType<typeof import("./actions")>;
