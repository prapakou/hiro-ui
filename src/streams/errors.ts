import { BehaviorSubject } from "rxjs";

import { createSubscribedState } from "./_helpers";

type ErrorState = Error | undefined;

const error$ = new BehaviorSubject<ErrorState>(undefined);

const setError = (e: Error) => {
  error$.next(e);
};

export const errorStore = {
  actions: {
    setError
  },
  getters: {
    useError: createSubscribedState<ErrorState>(error$)
  }
};
