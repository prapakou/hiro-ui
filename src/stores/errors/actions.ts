import { createStandardAction } from "typesafe-actions";

import { ErrorMessage } from "./constants";

export const errorSet = createStandardAction("ERROR_SET")<ErrorMessage>();
export const errorClear = createStandardAction("ERROR_CLEAR")();
