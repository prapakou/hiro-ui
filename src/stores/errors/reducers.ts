import { createReducer } from "typesafe-actions";

import { ErrorActionsType, ErrorStateType } from "./constants";

export const errorReducer = createReducer<ErrorStateType, ErrorActionsType>(
  null
)
  .handleAction("ERROR_SET", (_, { payload: { name = "Error", message } }) => ({
    name,
    message
  }))
  .handleAction("ERROR_CLEAR", () => null);
