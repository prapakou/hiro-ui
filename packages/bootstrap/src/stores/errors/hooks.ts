import { useCallback } from "react";
import { useMappedState, useDispatch } from "redux-react-hook";
import { get } from "lodash-es";

import { HIRO_NAMESPACE } from "../constants";

import { ERROR_NAMESPACE, ErrorMessage } from "./constants";
import { errorSet, errorClear } from "./actions";

export const useErrorState = () => {
  const mapState = useCallback(
    state => get(state, [HIRO_NAMESPACE, ERROR_NAMESPACE]),
    []
  );

  return useMappedState(mapState);
};

export const useErrorDispatch = () => {
  const dispatch = useDispatch();

  const setError = useCallback(
    (error: ErrorMessage) => dispatch(errorSet(error)),
    [dispatch]
  );
  const clearError = useCallback(() => dispatch(errorClear()), [dispatch]);

  return { setError, clearError };
};
