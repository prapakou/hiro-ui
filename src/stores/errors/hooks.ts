import { useCallback } from "react";
import { get } from "lodash-es";
import { HIRO_NAMESPACE } from "../constants";
import { ERROR_NAMESPACE, IErrorMessage } from "./constants";
import { useMappedState, useDispatch } from "redux-react-hook";
import { errorSet, errorClear } from "./actions";

export const useErrorState = () => {
  const mapState = useCallback(
    state => get(state, [HIRO_NAMESPACE, ERROR_NAMESPACE]),
    []
  );

  return useMappedState(mapState);
};

export const useErrorDispatch = () => {
  const dispatch = useDispatch();

  const setError = useCallback(
    (error: IErrorMessage) => dispatch(errorSet(error)),
    []
  );
  const clearError = useCallback(() => dispatch(errorClear()), []);

  return { setError, clearError };
};
