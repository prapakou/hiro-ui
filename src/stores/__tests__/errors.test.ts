import configureStore, { MockStoreEnhanced } from "redux-mock-store";

import { errorSet, errorClear, errorReducer } from "../errors";

const mockStore = configureStore();
let store: MockStoreEnhanced;

beforeEach(() => {
  store = mockStore({});
});

test("Error state", () => {
  let action: any = store.dispatch(errorSet({ message: "An error" }));
  expect(store.getActions()).toMatchSnapshot("Add");
  expect(errorReducer(undefined, action)).toMatchSnapshot("Add - Reducer");

  action = store.dispatch(errorClear());
  expect(store.getActions()).toMatchSnapshot("Clear");
  expect(errorReducer(undefined, action)).toMatchSnapshot("Clear - Reducer");
});
