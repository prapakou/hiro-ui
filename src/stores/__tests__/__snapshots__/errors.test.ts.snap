// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error state: Add - Reducer 1`] = `
Object {
  "message": "An error",
  "name": "Error",
}
`;

exports[`Error state: Add 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {
      "message": "An error",
    },
    "type": "ERROR_SET",
  },
]
`;

exports[`Error state: Clear - Reducer 1`] = `null`;

exports[`Error state: Clear 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {
      "message": "An error",
    },
    "type": "ERROR_SET",
  },
  Object {
    "meta": undefined,
    "payload": undefined,
    "type": "ERROR_CLEAR",
  },
]
`;
